index.ts

```diff
@@ -0,0 +1,154 @@
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in
+ *  all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+
+import * as path from 'path';
+import { clone, merge } from 'lodash';
+import * as typedoc from 'typedoc';
+import * as webpack from 'webpack';
+
+const PluginName = 'typedoc-webpack-plugin';
+
+const ColorReset = '\u001b[0m';
+const ColorYellow = '\u001b[33m';
+
+function isString(x: any): x is string {
+    return typeof x === 'string';
+}
+
+/**
+ * TypedocOptions is an Object that has either out or json.
+ */
+export class TypeDocOptions {
+    out?: string;
+    json?: string;
+    [key: string]: any
+}
+
+const DefaultTypeDocOptions: TypeDocOptions = {
+    out: './docs',
+    module: 'commonjs',
+    target: 'es5',
+    exclude: '**/node_modules/**/*.*',
+    experimentalDecorators: true,
+    excludeExternals: true
+};
+
+// This is a workaround against that `webpack.compilation.Compilation` in @types/webpack doesn't have fileTimestamps.
+interface Compilation extends webpack.compilation.Compilation {
+    fileTimestamps: Map<string, number>
+}
+
+/**
+ * Plugin class
+ */
+export class TypeDocWebpackPlugin {
+
+    private options: TypeDocOptions;
+    private readonly inputFiles: string | Array<string>;
+    
+    constructor(options: TypeDocOptions = DefaultTypeDocOptions, input: string | Array<string> = ['./'] ) {
+        
+        this.inputFiles = isString(input) ? [input] : input;
+    
+        // merge user options into default options and assign
+        this.options = merge(DefaultTypeDocOptions, options);
+    }
+    
+    /**
+     * @param {webpack.Compiler} compiler Webpack compiler object. @see <a href="https://webpack.github.io/docs/plugins.html#the-compiler-instance">label</a>
+     * @return void
+     */
+    apply(compiler: webpack.Compiler): void {
+        
+        const startTime = Date.now();
+        let prevTimestamps = new Map<string, number>();
+    
+        // If an absolute path is set in options.out or options.json, use that
+        // else if the output path is specified in webpack config, output typedocs are relative to that path
+        // else output typedocs are relative to the current working directory.
+        this.options = this.resolvedOptions(compiler);
+        
+        compiler.hooks.emit.tap(PluginName, (compilation: Compilation) => {
```

Comment:
As mentioned in #23 `Tapable.plugin` is deprecated.
Migration to new API `Compiler#hooks`.